virtcontainers/cgroups.go

```diff
@@ -1,197 +1,310 @@
 // Copyright (c) 2018 Huawei Corporation
+// Copyright (c) 2019 Intel Corporation
 //
 // SPDX-License-Identifier: Apache-2.0
 //
 
 package virtcontainers
 
 import (
-	"encoding/json"
+	"bufio"
 	"fmt"
+	"math"
+	"os"
+	"path/filepath"
+	"strings"
 
 	"github.com/containerd/cgroups"
-	"github.com/kata-containers/runtime/virtcontainers/pkg/annotations"
 	specs "github.com/opencontainers/runtime-spec/specs-go"
 )
 
-const (
-	vcpuGroupName       = "vcpu"
-	defaultCgroupParent = "/kata"
-)
-
-type sandboxCgroups struct {
-	commonParent cgroups.Cgroup
-	sandboxSub   cgroups.Cgroup
-	vcpuSub      cgroups.Cgroup
+type cgroupPather interface {
+	cgroups.Subsystem
+	Path(path string) string
 }
 
-func (s *Sandbox) newCgroups() error {
-	// New will still succeed when cgroup exists
-	// create common parent for all kata-containers
-	// e.g. /sys/fs/cgroup/cpu/vc
-	parent, err := cgroups.New(cgroups.V1,
-		cgroups.StaticPath(defaultCgroupParent), &specs.LinuxResources{})
-	if err != nil {
-		return fmt.Errorf("failed to create cgroup for %q", defaultCgroupParent)
-	}
+// unconstrained cgroups are placed here.
+// for example /sys/fs/cgroup/memory/kata/$CGPATH
+// where path is defined by the containers manager
+const cgroupKataPath = "/kata/"
+
+var cgroupsLoadFunc = cgroups.Load
+var cgroupsNewFunc = cgroups.New
 
-	// create sub-cgroup for each sandbox
-	// e.g. /sys/fs/cgroup/cpu/vc/<sandbox>
-	sandboxSub, err := parent.New(s.id, &specs.LinuxResources{})
+// V1Constraints returns the cgroups that are compatible with th VC architecture
+// and hypervisor, constraints can be applied to these cgroups.
+func V1Constraints() ([]cgroups.Subsystem, error) {
+	root, err := cgroupV1MountPoint()
 	if err != nil {
-		return fmt.Errorf("failed to create cgroup for %s/%s", defaultCgroupParent, s.id)
+		return nil, err
+	}
+	subsystems := []cgroups.Subsystem{
+		cgroups.NewCputset(root),
```

Comment:
Cputset or cpuset?

Reply:
Cputset is correct https://github.com/kata-containers/runtime/blob/master/vendor/github.com/containerd/cgroups/cpuset.go#L29

Reply:
I wonder this is a typo of `cgroups` package, I can't understand what this "t" means...